### ISSUE TITLE
ncurses fails to build on Cray system at LANL

### ISSUE THREAD

**Issue by jgalarowicz** [spack/spack#7315]
ncurses fails to build on a Cray platform at LANL.  It builds on the front-end node but not on the backend node (os=CNL).   I added the work around for pkg-config, but can't build ncurses on this Cray.
It fails in a similar way to the failure on the NERSC Cori Cray where the pkg-config work around was successful.  The work around doesn't appear to work on this Cray.

jegsgi/jeg_spack> **spack install ncurses os=CNL**
==> Installing pkg-config
==> Using cached archive: /usr/projects/packages/openspeedshop/jegsgi/jeg_spack/var/spack/cache/pkg-config/pkg-config-0.29.2.tar.gz
==> Staging archive: /usr/projects/packages/openspeedshop/jegsgi/jeg_spack/var/spack/stage/pkg-config-0.29.2-biyt4qbpfkxcr5s6jcmonmzd7pxvvagw/pkg-config-0.29.2.tar.gz
==> Created stage in /usr/projects/packages/openspeedshop/jegsgi/jeg_spack/var/spack/stage/pkg-config-0.29.2-biyt4qbpfkxcr5s6jcmonmzd7pxvvagw
==> No patches needed for pkg-config
==> Building pkg-config [AutotoolsPackage]
==> Executing phase: 'autoreconf'
==> Executing phase: 'configure'
==> Executing phase: 'build'
==> Executing phase: 'install'
==> Successfully installed pkg-config
  Fetch: 0.01s.  Build: 1m 9.26s.  Total: 1m 9.27s.
[+] /usr/projects/packages/openspeedshop/jegsgi/jeg_spack/opt/spack/cray-CNL-haswell/gcc-7.2.0/pkg-config-0.29.2-biyt4qbpfkxcr5s6jcmonmzd7pxvvagw
==> Installing ncurses
==> Error: SyntaxError: invalid syntax (<string>, line 486)
SyntaxError: SyntaxError: invalid syntax (<string>, line 486)

/usr/projects/packages/openspeedshop/jegsgi/jeg_spack/lib/spack/spack/build_environment.py:684, in child_process:
     27    
     28                # build up some context from the offending package so we can
     29                # show that, too.
  >> 30                package_context = get_package_context(tb)
     31    
     32                build_log = None
     33                if hasattr(pkg, 'log_path'):

jegsgi/jeg_spack> **cat ~/.spack/cray/packages.yaml** 
packages:
   bison:
      buildable: False
      paths:
         bison@2.7%gcc@4.9.3 arch=cray-CNL-haswell: /usr
   all:
      providers:
         pkgconfig: [pkg-config]
   mpich:
      buildable: False
      paths:
         mpich@7.6.2%gcc@4.9.3 arch=cray-CNL-haswell: /opt/cray/pe/mpt/7.6.2/gni//mpich-gnu/5.1

**Comment by jgalarowicz** [comment]
@mamelara @alalazo @skosukhin

I moved the new ncurses-6.1.tar.gz sources up to the Cray platform, but I don't think we are even getting to the source.  Something is happening in spack that prevents the build from even getting started.    Does anyone know where the syntax error lines are coming from - what source file in spack?

> spack install ncurses  os=CNL
==> Installing pkg-config
==> Fetching file:///usr/projects/packages/openspeedshop/jegsgi/jeg_spack/spack-compute-mirror/pkg-config/pkg-config-0.29.2.tar.gz
######################################################################## 100.0%
==> Staging archive: /usr/projects/packages/openspeedshop/jegsgi/jeg_spack/var/spack/stage/pkg-config-0.29.2-biyt4qbpfkxcr5s6jcmonmzd7pxvvagw/pkg-config-0.29.2.tar.gz
==> Created stage in /usr/projects/packages/openspeedshop/jegsgi/jeg_spack/var/spack/stage/pkg-config-0.29.2-biyt4qbpfkxcr5s6jcmonmzd7pxvvagw
==> No patches needed for pkg-config
==> Building pkg-config [AutotoolsPackage]
==> Executing phase: 'autoreconf'
==> Executing phase: 'configure'
==> Executing phase: 'build'
==> Executing phase: 'install'
==> Successfully installed pkg-config
  Fetch: 0.06s.  Build: 1m 39.71s.  Total: 1m 39.77s.
[+] /usr/projects/packages/openspeedshop/jegsgi/jeg_spack/opt/spack/cray-CNL-haswell/gcc-7.2.0/pkg-config-0.29.2-biyt4qbpfkxcr5s6jcmonmzd7pxvvagw
==> Installing ncurses
==> Error: SyntaxError: invalid syntax (<string>, line 486)
SyntaxError: SyntaxError: invalid syntax (<string>, line 486)

/usr/projects/packages/openspeedshop/jegsgi/jeg_spack/lib/spack/spack/build_environment.py:684, in child_process:
     27    
     28                # build up some context from the offending package so we can
     29                # show that, too.
  >> 30                package_context = get_package_context(tb)
     31    
     32                build_log = None
     33                if hasattr(pkg, 'log_path'):


var/spack> lsr /tmp/jegsgi/spack-stage/spack-stage-*
/tmp/jegsgi/spack-stage/spack-stage-Fe2ACF:
total 8
4 drwx------ 2 jegsgi jegsgi 4096 Feb 22 09:38 .
4 drwxr-xr-x 4 jegsgi jegsgi 4096 Feb 22 09:42 ..

/tmp/jegsgi/spack-stage/spack-stage-_L6bZ4:
total 8
4 drwx------ 2 jegsgi jegsgi 4096 Feb 22 09:40 .
4 drwxr-xr-x 4 jegsgi jegsgi 4096 Feb 22 09:42 ..
var/spack> !!/*
lsr /tmp/jegsgi/spack-stage/spack-stage-*/*
ls: No match.
var/spack>

**Comment by jgalarowicz** [comment]
@mamelara  @alalazo @skosukhin

Do any of you know if spack is somehow limiting the memory being used when building for the Cray compute nodes?    I am seeing the WARNING messages below on the packages that fail to build on this Cray.   Right now those are: ncurses, libxml2, and now gettext.   These build for the login/front-end nodes.  I built ncurses and libxml2 manually and specified those install paths in ~/.spack/cray/packages.yaml to get farther.

==> Installing gettext
==> Fetching file:///usr/projects/packages/openspeedshop/jegsgi/jeg_spack/spack-compute-mirror/gettext/gettext-0.19.8.1.tar.xz
######################################################################## 100.0%
==> Staging archive: /usr/projects/packages/openspeedshop/jegsgi/jeg_spack/var/spack/stage/gettext-0.19.8.1-csnchyprfiec7ro6ay7cmrgti7e7ho6h/gettext-0.19.8.1.tar.xz
libhugetlbfs [tt-fey1:33958]: WARNING: Hugepage size 2097152 unavailablelibhugetlbfs [tt-fey1:33958]: WARNING: New heap segment map at 0x10000000000 failed: Cannot allocate memory
==> Error: ProcessError: Command exited with status -4:
    '/usr/projects/packages/openspeedshop/jegsgi/jeg_spack/opt/spack/cray-CNL-haswell/gcc-7.2.0/tar-1.29-am23fv4rzryn7rm7uberhysoyk4cv6nb/bin/tar' '-xf' '/usr/projects/packages/openspeedshop/jegsgi/jeg_spack/var/spack/stage/gettext-0.19.8.1-csnchyprfiec7ro6ay7cmrgti7e7ho6h/gettext-0.19.8.1.tar.xz'
==> Error: [Errno 2] No such file or directory: '/usr/projects/packages/openspeedshop/jegsgi/jeg_spack/var/spack/stage/gettext-0.19.8.1-csnchyprfiec7ro6ay7cmrgti7e7ho6h/spack-expanded-archive/spack-build.out'

**Comment by jgalarowicz** [comment]
@mamelara @alalazo @skosukhin

Hi all,
This module command (below) eliminates the problems building ncurses, libxml2, and gettext for the compute nodes on the LANL Cray.  So, I'm closing this issue.

module unload craype-hugepages2M

**Comment by mamelara** [comment]
Hmmm probably need to add a dummy package for craype-hugepages or something. I've built packages that need that module loaded into that environment and others that need it unloaded

**Comment by scheibelp** [comment]
Thanks for the update! A couple side notes:

> Does anyone know where the syntax error lines are coming from - what source file in spack?

If you run the spack install with `-d` like `spack -d install...` then it should provide more context about where the error occurred in spack.

Also - could you use ``` for formatting code blocks like in https://help.github.com/articles/basic-writing-and-formatting-syntax/#quoting-code? A few of the above traces are hard to follow.

**Comment by jgalarowicz** [comment]
@scheibelp Thanks for the tip about formatting.   I didn't know how to control that.   And the -d install tip.

@mamelara Would we then do: depends_on the craype-hugepages package, if we needed it loaded?

Thanks
Jim G
